### PR TITLE
Use cookie for authentication

### PR DESCRIPTION
**Pull request by Cruguah**
This pull request:

adds issue_token / cookies authentication + tests
fixes the is_online key

### CODE REVIEW

**Comment by iMicknl** [comment]
@Cruguah one of the main reasons why this PR didn't get merged yet (even as pre release / test) is because this will break the authentication for current users that are still using the auth token method. I am looking if we can combine this and only update the Config Flow for new users.

**Comment by iMicknl** [comment]
CI/CD failing on `R] [MANIFEST] Manifest keys have been sorted: domain, name, then alphabetical order`. I will fix this in `main` branch so you can just rebase. This is a new requirement from core.

File: .vscode/settings.json [hunk 1]
```diff
@@ -1,9 +1,30 @@
 {
```
Comment (iMicknl) [inline .vscode/settings.json:1]: Can you revert the changes here? They don't seem necessary for this PR.

File: README.md [hunk 2]
```diff
@@ -14,6 +14,7 @@ This integration will add the most important sensors of your Nest Protect device
 
 - Only Google Accounts are supported, there is no plan to support legacy Nest accounts
 - When Nest Protect (wired) occupancy is triggered, it will stay 'on' for 10 minutes. (API limitation)
+- Only *cookie authentication* is supported as Google removed the API key authentication method. This means that you need to login to the Nest website at least once to generate a cookie. This cookie will be used to authenticate with the Nest API. The cookie will be stored in the Home Assistant configuration folder and will be used for future requests.
```
Comment (iMicknl) [inline README.md:17]: Can we mention that when an user logs out / changes their password, that they need to reauthenticate? Is there a limit to the session currently, or will this work indefinitely?

File: custom_components/nest_protect/binary_sensor.py [hunk 3]
```diff
@@ -38,19 +38,19 @@ class NestProtectBinarySensorDescription(
         key="co_status",
         name="CO Status",
         device_class=BinarySensorDeviceClass.CO,
-        value_fn=lambda state: state == 3,
+        value_fn=lambda state: state != 0,
```
Comment (iMicknl) [inline custom_components/nest_protect/binary_sensor.py:41]: Can we move these changes to a separate PR?

File: custom_components/nest_protect/manifest.json [hunk 4]
```diff
@@ -1,18 +1,16 @@
...
\ No newline at end of file
+  "domain": "nest_protect",
+  "name": "Nest Protect",
+  "config_flow": true,
+  "documentation": "https://github.com/imicknl/ha-nest-protect",
+  "issue_tracker": "https://github.com/imicknl/ha-nest-protect/issues",
+  "requirements": [],
+  "codeowners": ["@imicknl"],
+  "iot_class": "cloud_polling",
+  "version": "0.4.0",
```
Comment (iMicknl) [inline custom_components/nest_protect/manifest.json:10]: Please don't bump the version number by hand in your PR.

File: custom_components/nest_protect/manifest.json [hunk 5]
```diff
@@ -1,18 +1,16 @@
 {
-    "domain": "nest_protect",
```
Comment (iMicknl) [inline custom_components/nest_protect/manifest.json:2]: Do we need the indentation changes in this file?

**Review by iMicknl — CHANGES_REQUESTED** [review]
Looks good. Some comments to get the review started. I will try to make some time to test this PR myself as well asap, so we can start merging this in.

File: .vscode/settings.json (hunk 1 above)
Comment (Cruguah) [reply .vscode/settings.json:1]: Done

File: README.md (hunk 2 above)
Comment (Cruguah) [reply README.md:17]: Done, It seems that the current implementation updates the uptime every 10 min or so, which will keep the session and cookie alive

File: custom_components/nest_protect/binary_sensor.py (hunk 3 above)
Comment (Cruguah) [reply custom_components/nest_protect/binary_sensor.py:41]: I've just copied the PR to keep it save and I'm not intedend to change it to much

File: custom_components/nest_protect/manifest.json (hunk 4 above)
Comment (Cruguah) [reply custom_components/nest_protect/manifest.json:10]: This make shure it's not interfering with the nest integration of home assistant.

File: custom_components/nest_protect/manifest.json (hunk 5 above)
Comment (Cruguah) [reply custom_components/nest_protect/manifest.json:2]: This is the default indentation of vsstudio and vscode for json files. Reverted to your indentation

File: custom_components/nest_protect/manifest.json (hunk 4 above)
Comment (iMicknl) [reply custom_components/nest_protect/manifest.json]: @Cruguah I know, but in this integration we keep our pre-releases in main (latest git version) and new releases on GitHub will bump the version number.

File: custom_components/nest_protect/binary_sensor.py (hunk 3 above)
Comment (iMicknl) [reply custom_components/nest_protect/binary_sensor.py:41]: @Cruguah ok! I misunderstood :).

In that case I will see if I can work on top of this myself.

Comment (m4v3r1ckNl) [reply custom_components/nest_protect/binary_sensor.py:41]: > @Cruguah ok! I misunderstood :).
> 
> In that case I will see if I can work on top of this myself.

Would love to see where this these two ends meet, holding off my update to 2023.3.1
Thanks for all your work guys.

File: custom_components/nest_protect/manifest.json (hunk 4 above)
Comment (Cruguah) [reply custom_components/nest_protect/manifest.json]: Removed the version number

**Comment by iMicknl** [comment]
Thanks a lot and good to see all checks are green. I will see if we can merge this in a beta branch, to keep this code in this repository as well. 

In order to be merged to main, the following issues would need to be fixed first: https://github.com/iMicknl/ha-nest-protect/labels/beta%20%28cookie-method%29

**Comment by Cruguah** [comment]
@iMicknl, I've implemented a reconfiguration option to renew the issue_token and cookie. Would you be so kind to update you HA instance with this version to see if that works?

You can reinstall the nest integration using HACS by reinstalling the current integration, Open the integration and click on:
image

Then you can find the (re)configure option in the integration:
image

Hope this helps!

**Comment by spyfly** [comment]
Would be nice if this could finally be merged so that cookie auth can be utilised.

**Comment by iMicknl** [comment]
@spyfly I would agree, but first all (bigger) issues should be resolved: https://github.com/iMicknl/ha-nest-protect/issues?q=is%3Aissue+is%3Aopen+label%3A%22beta+%28cookie-method%29%22.

**Comment by ricoroodenburg** [comment]
When can we expect a working version with cookie auth?

**Comment by iMicknl** [comment]
> When can we expect a working version with cookie auth? Also there is no documation available for configuring this integration. The integrations says: Check the README. But the README shows only how to install this integration but not how to configure.

We all work in our spare time (aka for free) on this integration, and my time is currently very limited. Happily some people in the community stood up and worked on integrating the cookie auth, but this PR is not yet ready to be merged into the main integration.

**Comment by Cruguah** [comment]
> > > When can we expect a working version with cookie auth? Also there is no documation available for configuring this integration. The integrations says: Check the README. But the README shows only how to install this integration but not how to configure.
> > 
> > 
> > We all work in our spare time (aka for free) on this integration, and my time is currently very limited. Happily some people in the community stood up and worked on integrating the cookie auth, but this PR is not yet ready to be merged into the main integration.
> 
> I understand and really appreciate your effort, sadly I don't have any experience for programming in Python. I'll waiting.

And by the way, the integration works (for most of us), but also logs some warning in your logfile. We definitly need to fix this, but it all works for now. As soon as I have some time I'll dig in the async_forward_entry_setups to keep the integration working.

**Comment by wafliron** [comment]
Can anyone confirm if the cookie with version keeps working in 2023.04 and beyond? I’m not clear if the warning about it showing up in 2023.03 is terminal or not (and haven’t upgraded past 2023.03 yet).

Thanks.

**Comment by Cruguah** [comment]
All issues have been resolved, but due to the lack of version control on this PR, no one knows which issue has been resolved because no one knows which version they are using.
And all reported issues were reported in an earlier version.
So please give this PR version a version number, or merge this PR into the main branch and start working on real problems based on versions.

PS. wafliron, if you are using the latest version (and you need to check this manualy) it will continue to work.

**Comment by iMicknl** [comment]
> So please give this PR version a version number, or merge this PR into the main branch and start working on real problems based on version

If someone is committed to fix the remaining issues, I am happy to merge this and release this as a pre-release.

**Comment by wafliron** [comment]
> PS. wafliron, if you are using the latest version (and you need to check this manualy) it will continue to work.

Thanks @Cruguah. 

Must admit I'm confused by the versioning/forks of this PR/integration (and confused about correct terminology!), but for the record I'm currently using the fork at https://github.com/nicoinch/ha-nest-protect (re-installed it 1-2 weeks ago).

For the benefit of anyone else reading this now/in the future: the above fork continued to work (with cookie auth) after upgrading past 2023.02 (to 2023.04.05). The warning "Detected integration that called async_setup_platforms instead of awaiting async_forward_entry_setups; this will fail in version 2023.3" persists in the logs, but it doesn't appear to have actually failed / broken anything obvious (confirmed at least the occupancy sensors from my Nest Protects continue to update properly in HA).

File: custom_components/nest_protect/config_flow.py [hunk 6]
```diff
@@ -1,26 +1,30 @@
 """Adds config flow for Nest Protect."""
 from __future__ import annotations
-
 from typing import Any, cast
```
Comment (iMicknl) [inline custom_components/nest_protect/config_flow.py:3]:
```suggestion
from typing import Any, cast

```

File: custom_components/nest_protect/config_flow.py [hunk 7]
```diff
@@ -1,26 +1,30 @@
...
+    CONF_COOKIES,
+    CONF_ISSUE_TOKEN,
+    CONF_REFRESH_TOKEN,
+    DOMAIN,
+    LOGGER,
+)
 from .pynest.client import NestClient
 from .pynest.const import NEST_ENVIRONMENTS
 from .pynest.exceptions import BadCredentialsException
+import voluptuous as vol
```
Comment (iMicknl) [inline custom_components/nest_protect/config_flow.py:21]:
```suggestion
```

File: custom_components/nest_protect/config_flow.py [hunk 8]
```diff
@@ -1,26 +1,30 @@
...
-
 from typing import Any, cast
-
 from aiohttp import ClientError
 from homeassistant import config_entries
 from homeassistant.config_entries import ConfigEntry
-from homeassistant.const import CONF_TOKEN, CONF_URL
+from homeassistant.core import callback
 from homeassistant.data_entry_flow import FlowResult
 from homeassistant.helpers.aiohttp_client import async_get_clientsession
```
Comment (iMicknl) [inline custom_components/nest_protect/config_flow.py:9]:
```suggestion
from homeassistant.helpers.aiohttp_client import async_get_clientsession
import voluptuous as vol

```

**Review by iMicknl — COMMENTED** [review]
I can't edit your branch unfortunately, so trying to fix your linting issues via interface..

File: .github/workflows/pre-commit.yml [hunk 9]
```diff
@@ -2,7 +2,7 @@ name: Linters (flake8, black, isort)
 
 on:
   pull_request:
-
+
```
Comment (iMicknl) [inline .github/workflows/pre-commit.yml:5]:
```suggestion
 
```

**Comment by iMicknl** [comment]
@Cruguah could you unprotect your main branch? You did your  PR from the main branch, and thus I cannot help you and make changes. My proposal would be to fix this PR, bring it to beta branch and release this one as prerelease, so that HACS users can easily install and we can do version management.

**Comment by Cruguah** [comment]
> @Cruguah could you unprotect your main branch? You did your PR from the main branch, and thus I cannot help you and make changes. My proposal would be to fix this PR, bring it to beta branch and release this one as prerelease, so that HACS users can easily install and we can do version management.

I just did it!

**Comment by iMicknl** [comment]
@Cruguah not yet "You can’t commit to main because it is a protected branch". You can override this as admin, but you would need to remove the protection (or do the PR from another branch), for me to be able to help.

**Comment by Cruguah** [comment]
Ok, this looks better!

**Comment by iMicknl** [comment]
@Cruguah see 
image

**Comment by Cruguah** [comment]
Yep, and installed this beta version.
I'll keep an eye on the logfile, but up until now, all looks good!

**Comment by iMicknl** [comment]
I tagged all issues with https://github.com/iMicknl/ha-nest-protect/issues?q=is%3Aissue+is%3Aopen+label%3A%22beta+%28cookie-method%29%22, so that we can easily track which issues we need to resolve to merge to main.

**Comment by wafliron** [comment]
Confirming that after upgrading to 0.4.0b1 my Nest Protects have continued to work fine in HA and the async_setup_platforms warning is gone. Thanks for your work on this @iMicknl and @Cruguah.

**Comment by m4v3r1ckNl** [comment]
My Nest Protect integration (2) were broken since 2023.6.x but now HA is on 2023.7.1. Is the use of the BETA 0.4.0b1 still valid for my version of HA?

**Comment by iMicknl** [comment]
@m4v3r1ckNl you are using the fork, not my version if it was broken. Remove the custom repository from HACS and install beta 0.4.0b2.

**Comment by m4v3r1ckNl** [comment]
Thank you @iMicknl for your reply, but I'm not sure which version (forked) I installed. I used your cookie auth method on my instance of Nest Protect. Please see my screencap. All was functioning 100%.

Screenshot 2023-07-11 at 13 30 24

**Comment by iMicknl** [comment]
@m4v3r1ckNl please create a new issue if you face issues. However, if you follow my instructions (remove the custom repository), it should be fixed.

**Comment by m4v3r1ckNl** [comment]
Thank you @iMicknl I'll give that a try, the current integration is broken anyway. Cheers

**Comment by m4v3r1ckNl** [comment]
Hi @iMicknl 

Up-and-running again. Deleted the nest_protect under custom components and reinstalled the beta 0.4.0b2 today.

Screenshot 2023-08-01 at 12 02 26

Screenshot 2023-08-01 at 11 28 32

Thank you for your help!
